scripts/aks-with-registry.sh

```diff
@@ -0,0 +1,260 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit # exit immediately if a command exits with a non-zero status.
+set -o nounset # exit when script tries to use undeclared variables.
+set -o pipefail # make the pipeline fail if any command in it fails.
+
+REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+# shellcheck source=hack/ensure-azcli.sh
+source "${REPO_ROOT}/hack/ensure-azcli.sh" # install az cli and login using WI
+# shellcheck source=hack/ensure-tags.sh
+source "${REPO_ROOT}/hack/ensure-tags.sh" # set the right timestamp and job name
+
+
+main() {
+  # aks defaults
+  export MGMT_CLUSTER_NAME="${MGMT_CLUSTER_NAME:-mgmt-cluster}-${RANDOM_SUFFIX}" # management cluster name
+  export WRKLD_CLUSTER_NAME="${CLUSTER_NAME:-capz-aks}-${RANDOM_SUFFIX}" # workload cluster name
+  export AZURE_RESOURCE_GROUP="${AZURE_RESOURCE_GROUP:-capz-rg}-${RANDOM_SUFFIX}" # resource group name
+  export KUBERNETES_VERSION="${KUBERNETES_VERSION:-v1.30.2}"
+  export AZURE_LOCATION="${AZURE_LOCATION:-westus2}"
+  export AZURE_CONTROL_PLANE_MACHINE_TYPE="${AZURE_CONTROL_PLANE_MACHINE_TYPE:-"Standard_D2s_v3"}"
+  export AZURE_NODE_MACHINE_TYPE=${AZURE_NODE_MACHINE_TYPE:-"Standard_D2s_v3"}
+  export AZURE_CONTROL_PLANE_NODE_COUNT="${AZURE_CONTROL_PLANE_NODE_COUNT:-1}"
+  export MGMT_CLUSTER_KUBECONFIG="mgmt-kubeconfig.config"
+  export AZWI_STORAGE_ACCOUNT="${AZWI_STORAGE_ACCOUNT:-capzwi}-${RANDOM_SUFFIX}"
+  export AZWI_STORAGE_CONTAINER="${AZWI_STORAGE_CONTAINER:-capzwi-container}-${RANDOM_SUFFIX}"
+
+
+  export AZURE_SUBSCRIPTION_ID="${AZURE_SUBSCRIPTION_ID:-}"
+  export AZURE_CLIENT_ID="${AZURE_CLIENT_ID:-}"
+  # export AZURE_CLIENT_SECRET="${AZURE_CLIENT_SECRET:-}" this script is designed to use WI
+  export AZURE_TENANT_ID="${AZURE_TENANT_ID:-}"
+
+  # TODO: remove printing all the env variables
+  echo "--------------------------------"
+  echo "MGMT_CLUSTER_NAME: $MGMT_CLUSTER_NAME"
+  echo "WRKLD_CLUSTER_NAME: $WRKLD_CLUSTER_NAME"
+  echo "AZURE_RESOURCE_GROUP: $AZURE_RESOURCE_GROUP"
+  echo "KUBERNETES_VERSION: $KUBERNETES_VERSION"
+  echo "AZURE_LOCATION: $AZURE_LOCATION"
+  echo "AZURE_CONTROL_PLANE_MACHINE_TYPE: $AZURE_CONTROL_PLANE_MACHINE_TYPE"
+  echo "AZURE_NODE_MACHINE_TYPE: $AZURE_NODE_MACHINE_TYPE"
+  echo "AZURE_CONTROL_PLANE_NODE_COUNT: $AZURE_CONTROL_PLANE_NODE_COUNT"
+  echo "MGMT_CLUSTER_KUBECONFIG: $MGMT_CLUSTER_KUBECONFIG"
+  echo "AZWI_STORAGE_ACCOUNT: $AZWI_STORAGE_ACCOUNT"
+  echo "AZWI_STORAGE_CONTAINER: $AZWI_STORAGE_CONTAINER"
+
+  echo "AZURE_SUBSCRIPTION_ID: $AZURE_SUBSCRIPTION_ID"
+  echo "AZURE_CLIENT_ID: $AZURE_CLIENT_ID"
+  echo "AZURE_TENANT_ID: $AZURE_TENANT_ID"
+  echo "--------------------------------"
+
+
+  echo "creating resource group and aks cluster"
+  create_resource_group_and_aks
+
+  unset AZURE_STORAGE_KEY
+  unset AZURE_STORAGE_ACCOUNT
+
+  echo "creating storage account ${AZWI_STORAGE_ACCOUNT}"
```

Comment:
The main idea behind using AKS here is to avoid needing to create and manage keys and a storage account ourselves. Instead we should be passing `--enable-oidc-issuer` to the `az aks create` command which will do all of this automatically.

Reply:
we shouldn't need `--enable-oidc-issuer` either, since the vm is running in Azure we should be able to use the Managed Identity to make calls to azure, assuming we've set up RBAC roles required on the MI

Reply:
Is `Contributor` role for the sub a good enough RBAC on the MI ?

Reply:
Generally yes, but we specify a couple extra ones:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/15f32c998a3a208a0eb4c55ca4e5e39573e31074/scripts/kind-with-registry.sh#L162-L170

RBAC Administrator is to create role assignments in cluster-autoscaler tests which set up their own workload ID with an AKS cluster like we do here. Storage Blob Data Reader lets the `ci-version` and `custom-builds` tests read custom built k8s artifacts from a storage account. Outside of those cases, Contributor is enough.